File: src/lib/database/new.ts
import { invoke as i } from '@tauri-apps/api/tauri';
import { handle_error } from './';
import type { Team, Event, Participant, NewTeam, NewEvent, NewParticipant } from './';

export async function new_team(team: NewTeam) {
	const { name, individual } = team;

	try {
		const team = await i<Team>('new_team', { name, individual });
		alert(`Team "${team.name}" added successfully!`);
		return team;
	} catch (e) {
		handle_error(e as Record<string, string>);
	}
}

export async function new_participant(participant: NewParticipant, team_id: number) {
	const { first_name, last_name } = participant;

	try {
		const participant = await i<Participant>('new_participant', { first_name, last_name, team_id });
		alert(`Participant "${participant.first_name} ${participant.last_name}" added scucessfully!`);
		return participant;
	} catch (e) {
		handle_error(e as Record<string, string>);
	}
}

export async function new_event(event: NewEvent) {
	const { name, event_type } = event;

	try {
		const db_event = await i<Event>('new_event', { name, event_type });
		alert(`Event "${db_event.name}" added successfully!`);
		return db_event;
	} catch (e) {
		handle_error(e as Record<string, string>);
	}
}

export async function enroll_team_in_events(team_id: number, events: number[] | number) {
	events = (Array.isArray(events) ? events : [events]) as number[];

	const promises = [];

	for (const event_id of events) {
		promises.push(i('enroll_team_in_events', { team_id, event_id }));
	}

	return Promise.all(promises);
}

File: src-tauri/src/commands/new.rs
use crate::db::structs::{Event, EventEntry, Participant, Team};
use crate::error::Error;
use crate::utils::connect_to_db;
use tauri::AppHandle;

#[tauri::command(rename_all = "snake_case")]
pub fn new_team(app: AppHandle, name: &str, individual: bool) -> core::result::Result<Team, Error> {
    let database = connect_to_db(app);
    let new_team = database.new_team(name, individual)?;

    Ok(new_team)
}

#[tauri::command(rename_all = "snake_case")]
pub fn new_participant(
    app: AppHandle,
    first_name: &str,
    last_name: &str,
    team_id: i32,
) -> core::result::Result<Participant, String> {
    let database = connect_to_db(app);
    let new_participant = database
        .new_participant(first_name, last_name, team_id)
        .map_err(|e| e.to_string())?;

    Ok(new_participant)
}

#[tauri::command(rename_all = "snake_case")]
pub fn new_event(
    app: AppHandle,
    name: &str,
    event_type: &str,
) -> core::result::Result<Event, String> {
    let database = connect_to_db(app);
    let new_event = database
        .new_event(name, event_type)
        .map_err(|e| e.to_string())?;

    Ok(new_event)
}

#[tauri::command(rename_all = "snake_case")]
pub fn enroll_team_in_events(
    app: AppHandle,
    team_id: i32,
    event_id: i32,
) -> core::result::Result<EventEntry, String> {
    let database = connect_to_db(app);
    let new_event_entry = database
        .new_event_entry(team_id, event_id)
        .map_err(|e| e.to_string())?;

    Ok(new_event_entry)
}
